app/javascript/components/workflows/professor/exams/new/editor/components/questions/Code.tsx

```diff
@@ -32,159 +229,133 @@ interface LockStateInfo {
+  const isFile = value && 'file' in value;
+  const isText = value && 'text' in value;
+  const isNone = !value;
+  const { files } = useContext(ExamContext);
+  const first = firstFile(files);
+  return (
+    <>
+      <ButtonGroup>
+        <Button
+          variant={isNone ? 'secondary' : 'outline-secondary'}
+          active={isNone}
+          onClick={(): void => {
+            onChange(null);
+          }}
+        >
+          None
+        </Button>
+        <Button
+          variant={isText ? 'secondary' : 'outline-secondary'}
+          active={isText}
+          onClick={(): void => {
+            onChange({ text: '', marks: [] });
+          }}
+        >
+          Supply code
+        </Button>
+        <Button
+          variant={isFile ? 'secondary' : 'outline-secondary'}
+          active={isFile}
+          onClick={(): void => {
+            onChange({ file: first.relPath });
+          }}
+        >
+          Choose a file
+        </Button>
+      </ButtonGroup>
+      {isText && (
+        <Fields
+          names={['lang', 'initial']}
+          component={EditSuppliedCode}
+          controlId={controlId}
+        />
+      )}
+      {value && 'file' in value && (
+        <>
+          <FilePickerSelectWithPreview
+            options={files}
+            selected={[{
+              type: 'file',
+              path: value.file,
+            }]}
+            onChange={(arr): void => {
+              const lastSelected = arr[arr.length - 1];
+              if (!lastSelected) return;
+              if (lastSelected.type === 'dir') return;
+              onChange({ file: lastSelected.path });
+            }}
+          />
+        </>
+      )}
+    </>
+  );
+};
+
 const Code: React.FC<CodeProps> = (props) => {
   const {
-    info,
     qnum,
     pnum,
     bnum,
-    value,
-    onChange,
   } = props;
-  const { prompt, lang, initial } = info;
-  const { fmap } = useContext(ExamContext);
-  const f = fmap[initial];
-  if (f?.filedir === 'dir') {
-    throw new Error('Code initial cannot be a directory.');
-  }
-  const text = value?.text ?? f?.contents ?? '';
-  const marks = value?.marks ?? f?.marks ?? [];
-  const [lockState, setLockState] = useState<LockStateInfo>({
-    enabled: false,
-    active: false,
-  });
-  let title: string;
-  if (lockState.enabled) {
-    if (lockState.active) {
-      title = 'Region locked; click to unlock';
-    } else {
-      title = 'Lock region';
-    }
-  } else {
-    title = 'No region selected to lock';
-  }
   return (
     <>
       <Prompted
         qnum={qnum}
         pnum={pnum}
         bnum={bnum}
-        prompt={prompt.value}
-        onChange={(newPrompt): void => {
-          onChange({
-            ...info,
-            prompt: {
-              type: 'HTML',
-              value: newPrompt,
-            },
-          }, value);
-        }}
       />
-      <Form.Group as={Row} controlId={`${qnum}-${pnum}-${bnum}-answer`}>
+      <Form.Group as={Row}>
```

Comment:
This change in placement of the `controlId` is unfortunate - it breaks some of the accessibility.  (Specifically, labels need to have a `for` field.)